### PR TITLE
bpo-42862: Use functools.lru_cache iso. _sqlite.Cache in sqlite3 module

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue42862

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @berkerpeksag, @rhettinger, @serhiy-storchaka.

**Comment by erlend-aasland** [comment]
Hm, the Windows tests fails at the same places as #23046 failed. I'm guessing this has something to do with the way I've implemented the GC protocol in `connection.c`? @serhiy-storchaka, @berkerpeksag could one of you point me in the right direction here?

**Comment by erlend-aasland** [comment]
Under the assumption that 6fd181180e1d748d1f0982821cc80f99c60c811a is acceptable, this PR is now ready for review, @berkerpeksag.

**Comment by erlend-aasland** [comment]
@rhettinger Would you mind reviewing the use of `functools.lru_cache` in this PR?

**Comment by erlend-aasland** [comment]
FYI, rebased onto master

**Comment by erlend-aasland** [comment]
FYI @berkerpeksag, I've rebased onto `master` again, bco. #24801.

**Comment by erlend-aasland** [comment]
FYI, synced with `master`.

**Comment by erlend-aasland** [comment]
Rebased onto master to resolve conflicts. Also added a tiny optimisation: increase the default cache size from 100 to 128. I'm guessing it is way to late to get this into 3.10 now :)

**Comment by erlend-aasland** [comment]
@pablogsal, two questions:

1. There's some issues with the Windows tests. I've added a workaround in f1f023bb14d0db88611522352cf93866f1f8e92e, but I don't understand why this is needed. One possible reason could be that SQLite is holding on to `TESTFN` after the test is finished. I haven't devoted too much time to investigate this yet.
2. Is it ok to request Raymonds review here? `functools` is his domain.

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst]: What's the reason for adding a public API? Previously, statement caching was an implementation detail apart from `cached_statements` of `sqlite3.connect()` (which IMO was a bad idea)

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I used it for debugging and felt it was a nice feature (and it was very convenient to implement, since we're now using functools). Looking at it now, I see no reason to add this to the API. I'll remove it. As you say, caching is an implementation detail.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Fixed in 55342a89168a48bbc14f0a314c801509e453eff1. I see no way to explicitly test the caching now, so I removed the tests.

**Comment by pablogsal** [comment]
> 2\. Is it ok to request Raymonds review here? `functools` is his domain.

Unless you are changing the module itself, is not very relevant, but you are always free to ask for his advice :)



> * There's some issues with the Windows tests. I've added a workaround in [f1f023b](https://github.com/python/cpython/commit/f1f023bb14d0db88611522352cf93866f1f8e92e), but I don't understand why this is needed. One possible reason could be that SQLite is holding on to `TESTFN` after the test is finished. I haven't devoted too much time to investigate this yet.

That seems like some genuine problem. Maybe @vstinner has seen this before?

File: Modules/_sqlite/connection.c
Comment (pablogsal) [inline Modules/_sqlite/connection.c]: Be careful with static globals as this is not subinterpreter-friendly. @vstinner or @ericsnowcurrently can explain further.

File: Modules/_sqlite/connection.c
Comment (pablogsal) [inline Modules/_sqlite/connection.c:62]: Check the return value of `PyLong_FromLong` and act accordingly

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Currently, there's no module state in sqlite3. Converting the sqlite3 state to heap types has been the first step on the path to sqlite3 multi-phase initialisation and module state. I can use this PR to establish a static global state, and put `_lru_cache` there.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:62]: Will do!

File: Modules/_sqlite/connection.c
Comment (pablogsal) [inline Modules/_sqlite/connection.c]:
```suggestion
```

File: Lib/sqlite3/test/hooks.py
Comment (pablogsal) [inline Lib/sqlite3/test/hooks.py]: Could you add a small comment why we need to explicitly call `gc_collect()`?

**Comment by pablogsal** [comment]
When you are done with the fixes, could you test with the buildbots using the buildbot label?

File: Lib/sqlite3/test/hooks.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/hooks.py]: Yes, I actually had that in mind.

**Comment by erlend-aasland** [comment]
> When you are done with the fixes, could you test with the buildbots using the buildbot label?

Of course!

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I added a global state struct and put `_lru_cache` there. I can add the rest of the static data to it in a separate PR. One step closer to multi-phase init and sub-interpreter support.

See bb186dca837c805cf5d78ae7675d8a43a0b217cd

**Comment by erlend-aasland** [comment]
The buildbots seems to have revealed some GC issues. I've marked this as draft until I've figured out what the problem is.

**Comment by erlend-aasland** [comment]
Merged `main` into this PR, to resolve https://github.com/python/cpython/pull/24203#issuecomment-845025674

**Comment by erlend-aasland** [comment]
On hold until #26104 is merged.

**Comment by erlend-aasland** [comment]
I messed up the previous merge so I had to rebase and force push. Sorry 'bout the mess.

File: Lib/sqlite3/test/dbapi.py
Comment (pablogsal) [inline Lib/sqlite3/test/dbapi.py]: As @vstinner I think this is wrong: if this is really needed, is hiding a real bug

**Review by pablogsal — CHANGES_REQUESTED** [review]

File: Lib/sqlite3/test/dbapi.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: Yes, I agree. Let's try to fix this in #26462 and then rebase this PR onto `main` when it is resolved. For now, I'll revert the workaround.

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
@erlend-aasland is this still a draft (I assume not)? :)

**Comment by erlend-aasland** [comment]
> is this still a draft (I assume not)? :)

Nope, I was waiting for the CI, just to be sure :)

**Comment by pablogsal** [comment]
Great job @erlend-aasland! :tada:

**Comment by erlend-aasland** [comment]
Thanks, Pablo! 😃

**Pull request merged**
